numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -0,0 +1,268 @@
+/* Array Descr Object */
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include "structmember.h"
+#include "assert.h"
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#include <numpy/ndarraytypes.h>
+#include "npy_pycompat.h"
+
+#include "dtypemeta.h"
+
+
+static void
+dtypemeta_dealloc(PyArray_DTypeMeta *self) {
+    /*
+     * PyType_Type asserts Py_TPFLAGS_HEAPTYPE as well. Do not rely on
+     * a python debug build though.
+     */
+    assert(((PyTypeObject *)self)->tp_flags & Py_TPFLAGS_HEAPTYPE);
+    Py_XDECREF(self->scalar_type);
+    PyType_Type.tp_dealloc((PyObject *) self);
+}
+
+static PyObject *
+dtypemeta_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyErr_SetString(PyExc_TypeError,
+            "Preliminary-API: Cannot subclass DType.");
+    return NULL;
+}
+
+static int
+dtypemeta_init(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyErr_SetString(PyExc_TypeError,
+            "Preliminary-API: Cannot initialize DType class.");
+    return -1;
+}
+
+/**
+ * tp_is_gc slot of Python types. This is implemented only for documentation
+ * purposes to indicate and document the subtleties involved.
+ *
+ * Python Type objects are either statically created (typical C-Extension type)
+ * or HeapTypes (typically created in Python).
+ * HeapTypes have the Py_TPFLAGS_HEAPTYPE flag, and are garbage collected, our
+ * DTypeMeta instances (`np.dtype` and its subclasses) *may* be HeapTypes
+ * if the Py_TPFLAGS_HEAPTYPE flag is set (they are created from Python).
+ * They are not for legacy DTypes or np.dtype itself.
+ *
+ * @param self
+ * @return nonzero if the object is garbage collected
+ */
+static NPY_INLINE int
+dtypemeta_is_gc(PyObject *dtype_class)
+{
+    return PyType_Type.tp_is_gc(dtype_class);
+}
+
+
+static int
+dtypemeta_traverse(PyArray_DTypeMeta *type, visitproc visit, void *arg)
+{
+    /*
+     * We have to traverse the base class (if it is a HeapType).
+     * PyType_Type will handle this logic for us.
+     * This function is currently not used, but may be necessary in the future
+     * when we implement HeapTypes (python/dynamically defined types).
+     */
+    assert(!type->is_legacy && (PyTypeObject *)type != &PyArrayDescr_Type);
+    Py_VISIT(type->singleton);
+    Py_VISIT(type->scalar_type);
+    return PyType_Type.tp_traverse((PyObject *)type, visit, arg);
+}
+
+
+static PyObject *
+legacy_dtype_default_new(PyArray_DTypeMeta *self,
+        PyObject *args, PyObject *kwargs)
+{
+    /* TODO: This should allow endianess and possibly metadata */
```

Comment:
I am happy to just disable this completely for now as well. Or, put a PreliminaryAPI warning here. Which I think would be nice in general, but I am not sure I care too much about right away (it is not like this is a tricky API choice), that `type(instance)()` gives the default instance.

The main annoyance about this PR is, how to create the TypeStruct in a way that is hidden from the public.

Reply:
Warnings are probably annoying, given that you'll then have to filter them out again if you want to use this internally (applies to any preliminary API). How about just putting a big warning in the docs for any such object? And on the Python side, keep it out of the main namespace.

Reply:
True, as this is the python side, for a long time it probably only matters in tests though. `np.dtype` does not really have a test right now. I could change the repr away from `np.dtype[float64]` if you prefer. E.g. just add an underscore to make it "private".